sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc

```diff
@@ -0,0 +1,302 @@
+      
+      DaqDecoderSBNDPMT & operator = (DaqDecoderSBNDPMT const &) = delete;
+      
+      DaqDecoderSBNDPMT & operator = (DaqDecoderSBNDPMT &&) = delete;
+
+      std::vector<art::Handle<artdaq::Fragments>> readHandles( art::Event const & event ) const;
+
+      artdaq::Fragments readFragments( std::vector<art::Handle<artdaq::Fragments>> handles );
+
+      void processFragment( const artdaq::Fragment &artdaqFragment );
+
+      void produce(art::Event & e) override;
+
+    private:
+
+      template <std::size_t NBits, typename T>
+        static constexpr std::pair<std::array<std::size_t, NBits>, std::size_t>
+          setBitIndices(T value) noexcept;
+      
+      std::vector<art::InputTag> m_input_tags;
+
+      using OpDetWaveformCollection    = std::vector<raw::OpDetWaveform>;
+      using OpDetWaveformCollectionPtr = std::unique_ptr<OpDetWaveformCollection>;
+      OpDetWaveformCollectionPtr fOpDetWaveformCollection;  
+    
+      using PMTDigitizerInfoCollection    = std::vector<pmtAnalysis::PMTDigitizerInfo>;
+      using PMTDigitizerInfoCollectionPtr = std::unique_ptr<PMTDigitizerInfoCollection>;
+      PMTDigitizerInfoCollectionPtr fPMTDigitizerInfoCollection;  
+
+      // Association
+
+  };
+
+}
+
+
+
+daq::DaqDecoderSBNDPMT::DaqDecoderSBNDPMT(Parameters const & params)
+  : art::EDProducer(params)
+  , m_input_tags{ params().FragmentsLabels() }
+
+{
+  std::cout<<"  MessageFran: Initialized decoder\n";
+  // Output data products
+  produces<std::vector<raw::OpDetWaveform>>();
+  produces<std::vector<pmtAnalysis::PMTDigitizerInfo>>();
+
+}
+
+
+template <std::size_t NBits, typename T>
+constexpr std::pair<std::array<std::size_t, NBits>, std::size_t>
+daq::DaqDecoderSBNDPMT::setBitIndices(T value) noexcept {
+  
+  std::pair<std::array<std::size_t, NBits>, std::size_t> res;
+  auto& [ indices, nSetBits ] = res;
+  for (std::size_t& index: indices) {
+    index = (value & 1)? nSetBits++: NBits;
+    value >>= 1;
+  } // for
+  return res;
+  
+} 
+
+
+std::vector<art::Handle<artdaq::Fragments>> daq::DaqDecoderSBNDPMT::readHandles( art::Event const & event ) const{
+
+  // Normally or the CAENV1730 or the ContainerCAENV1730 are full
+  // We return all the non-empty ones 
+
+  std::vector<art::Handle<artdaq::Fragments>> handles;
+  art::InputTag this_input_tag; 
+
+  for( art::InputTag const& input_tag : m_input_tags ) {
+
+    art::Handle<artdaq::Fragments> thisHandle;
+
+    if( !event.getByLabel<std::vector<artdaq::Fragment>>(input_tag, thisHandle) ) continue;
+    if( !thisHandle.isValid() || thisHandle->empty() ) continue;
+
+    handles.push_back( thisHandle );
+
+  }
+
+  return handles;
+
+} 
+
+
+artdaq::Fragments daq::DaqDecoderSBNDPMT::readFragments( std::vector<art::Handle<artdaq::Fragments>> handles ) {
+
+  // Hopefully-not-too-sloppy-code to create the fragment list out of the container  
+  artdaq::Fragments fragments;
+  
+  for( const auto& handle : handles ) {
+    
```

Comment:
Suggestion: this code assumes that each handle has at least one fragment in it; this is guaranteed by the caller, so there is no actual problem, but to document that this is a requirement and to have a check on it we can add an assertion:
```suggestion
    assert(!handle->empty());
```
(it needs to `#include <cassert>`)

Reply:
f777f67